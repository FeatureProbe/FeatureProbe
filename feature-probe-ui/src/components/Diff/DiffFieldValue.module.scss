.diff-field-value {
  display: inline-block;
  padding: 0 4px;
  font-family: PingFangSC-Regular;
  font-size: 12px;
  color: #495057;
  letter-spacing: 0;
  line-height: 18px;
  font-weight: 400;
  height: 18px;
  border-radius: 4px;
  white-space: nowrap;

  &.field-value-remove {
    background-color: #feecef !important;
  }

  &.field-value-add {
    background-color: #e5fbf0 !important;
  }

  &:nth-child(n + 2) {
    margin-left: 8px;
  }
}
