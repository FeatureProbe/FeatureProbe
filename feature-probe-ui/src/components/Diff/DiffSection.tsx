import { cloneDeep } from 'lodash';
import { ReactNode, useMemo } from 'react';
import idiff, { DiffParam, DiffResult } from './diff';
import styles from './DiffSection.module.scss';

interface DiffSectionProps {
  title: string;
  before: unknown;
  after: unknown;
  renderContent: (diffContent: DiffResult) => ReactNode;
  beforeDiff?: (before: unknown, after: unknown) => unknown[];
}

const DiffSection: React.FC<DiffSectionProps> = (props) => {
  const { title, before, after, renderContent, beforeDiff } = props;

  const diffContent = useMemo(() => {
    let left, right;
    if (beforeDiff) {
      [left, right] = beforeDiff(cloneDeep(before), cloneDeep(after));
    } else {
      [left, right] = [before, after];
    }

    const diffContent = idiff(left as DiffParam, right as DiffParam);

    return diffContent;
  }, [before, after, beforeDiff]);

  let hide = false;
  if (diffContent.length === 1 && !diffContent[0].removed && !diffContent[0].modified && !diffContent[0].added) {
    hide = true;
  }

  return (
    <div hidden={hide} className={styles['diff-section']}>
      <div className={styles['head-line']}>
        <div className={styles['before']}>
          <div className={styles.title}>{title}</div>
        </div>
        <div className={styles['after']}>
          <div className={styles.title}>{title}</div>
        </div>
      </div>
      <div className={styles.content}>{renderContent(diffContent)}</div>
    </div>
  );
};

export default DiffSection;
