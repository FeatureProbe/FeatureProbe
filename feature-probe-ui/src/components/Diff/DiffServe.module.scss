.serve-diff {
  display: flex;
  background-color: #f8f9fa;

  & > * {
    width: 100%;
    margin-left: 24px;
    overflow-x: auto;
    display: flex;
    align-items: center;
    background-color: #fff;
    
    & > * {
        margin-right: 12px;
    }
  }
}
