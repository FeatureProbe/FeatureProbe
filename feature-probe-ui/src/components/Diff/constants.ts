export type diffType = 'remove' | 'modify' | 'add' | 'same';
export type positionType = 'before' | 'after';