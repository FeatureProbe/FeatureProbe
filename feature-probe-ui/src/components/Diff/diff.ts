import {diffArrays, ArrayChange} from 'diff';

const diff = {
  diffArrays,
};

export interface DiffObj {
  [x: string]: unknown;
}

export interface ArrayObj {
  __key?: string;
  __value?: unknown;
}

export type DiffParam = ArrayObj[] | DiffObj | unknown[]

export type DiffResult = (ArrayChange<unknown> & { modified?: boolean  })[];

const compared = (obj1: DiffObj | ArrayObj, obj2: DiffObj | ArrayObj) => {
  if (typeof obj1 === 'object' && typeof obj2 === 'object') {
    if (obj1.__key !== undefined && obj2.__key !== undefined) {
      if (typeof obj1.__value === 'object' && typeof obj2.__value === 'object' && obj1.__value && obj2.__value) {
        return diffObj(obj1.__value as DiffObj, obj2.__value as DiffObj).length === 1;
      }
      return obj1.__value === obj2.__value;
    }
    return diffObj(obj1 as ArrayObj[], obj2 as ArrayObj[]).length === 1;
  } else {
    return obj1 === obj2;
  }
};

function diffObj(left: DiffParam, right: DiffParam): ArrayChange<ArrayObj | DiffObj>[] {
  if (left instanceof Array && right instanceof Array) {
    return diff.diffArrays<ArrayObj, ArrayObj>(left as ArrayObj[], right as ArrayObj[], {
      comparator: compared,
    });
  } else {
    const leftArr = Object.keys(left).map((key) => {
      return { __key: key, __value: (left as DiffObj)[key] };
    }).sort((a, b) => {
      return a.__key > b.__key ? 1 : -1;
    });
    const rightArr = Object.keys(right).map((key) => {
      return { __key: key, __value: (right as DiffObj)[key] };
    }).sort((a, b) => {
      return a.__key > b.__key ? 1 : -1;
    });

    return diff.diffArrays(leftArr, rightArr, {
      comparator: compared,
    });
  }
}

const isTryToDiff = (value: ArrayObj) => {
  if (typeof value === 'object') {
    if (value.__key !== undefined && value.__value !== undefined) {
      if (typeof value.__value === 'object') {
        return {
          flag: true,
          type: 'key-value',
        };
      }
    } else {
      return {
        flag: true,
        type: 'object',
      };
    }
  }
  return {
    flag: false,
  };
};

const resetDiffResult = (diffContent: ArrayChange<DiffObj | ArrayObj>[]) => {
  let result: DiffResult = [];

  for (let i = 0; i < diffContent.length; i++) {
    if (diffContent[i].removed) {
      if (i + 1 < diffContent.length && diffContent[i + 1].added) {
        result = [...result, ...toModified(diffContent[i].value as DiffObj[], diffContent[i + 1].value as DiffObj[])];
        i++;
      } else {
        result.push(diffContent[i]);
      }
    } else {
      result.push(diffContent[i]);
    }
  }

  return result;
};

function toModified(removed: DiffObj[], added: DiffObj[]) {
  const result: DiffResult  = [];
  let modifies = [];
  let removeds = [];
  let addeds = [];
  let i = 0;
  for (i; i < removed.length; i++) {
    if (i >= added.length) {
      break;
    }
    const left = removed[i];
    const right = added[i];

    const isDiff = isTryToDiff(right);
    if (isDiff.flag) {
      if (removeds.length) {
        result.push({ count: removeds.length, removed: true, value: removeds });
        result.push({ count: addeds.length, added: true, value: addeds });
        removeds = [];
        addeds = [];
      }

      if (isDiff.type === 'key-value' && left.__key === right.__key) {
        modifies.push({ __key: right.__key, __value: resetDiffResult(diffObj(left.__value as DiffParam, right.__value as DiffParam)) });
      } else if(left.key !== right.__key) {
        removeds.push(left);
        addeds.push(right);
      } else {
        modifies.push({ count: 1, modified: true, value: resetDiffResult(diffObj(left, right)) });
      }
    } else {
      if (modifies.length) {
        result.push({ count: modifies.length, modified: true, value: modifies });
        modifies = [];
      }

      removeds.push(left);
      addeds.push(right);
    }
  }

  if (removeds.length) {
    result.push({ count: removeds.length, removed: true, value: removeds });
  }

  if (addeds.length) {
    result.push({ count: addeds.length, added: true, value: addeds });
  }

  if (modifies.length) {
    result.push({ count: modifies.length, modified: true, value: modifies });
  }

  const last = result[result.length - 1];
  if (removed.length > added.length) {
    if (last.modified) {
      result.push({ count: added.length - removed.length, removed: true, value: removed.slice(added.length) });
    } else if (last.added) {
      result[result.length - 2].value = result[result.length - 2].value.concat(removed.slice(added.length));
    }
  } else if (removed.length < added.length) {
    if (last.modified) {
      result.push({ count: removed.length - added.length, added: true, value: added.slice(removed.length) });
    } else if (last.added) {
      last.value = last.value.concat(added.slice(removed.length));
    }
  }

  return result;
};


function idiff (left: DiffParam, right: DiffParam): DiffResult {
  return resetDiffResult(diffObj(left, right));
}
export default idiff;
