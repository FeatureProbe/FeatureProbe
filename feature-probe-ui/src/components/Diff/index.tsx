import { useEffect, useRef } from 'react';
import Icon from 'components/Icon';
import { FormattedMessage } from 'react-intl';
import styles from'./index.module.scss';
import 'diff2html/bundles/css/diff2html.min.css';

interface DiffProps {
  content: string;
  height?: number;
  maxHeight?: number;
}

const Diff: React.FC<DiffProps> = (props) => {
  const ref = useRef<HTMLDivElement>(null);

  useEffect(() => {
    if (ref.current) {
      if (props.maxHeight && ref.current.clientHeight <= props.maxHeight) {
        ref.current.style.height = ref.current.clientHeight + 6 + 'px';
      }

      const ySides = ref.current.getElementsByClassName('d2h-file-side-diff');
      if (ySides[0] && ySides[1]) {
        ySides[0]?.addEventListener('scroll', function (e) {
          ySides[1].scrollTop = (e.target as HTMLDivElement).scrollTop;
        });
        ySides[1]?.addEventListener('scroll', function (e) {
          ySides[0].scrollTop = (e.target as HTMLDivElement).scrollTop;
        });
      }

      const xSides = ref.current.getElementsByClassName('d2h-code-wrapper');

      if (xSides[1].scrollWidth > xSides[1].clientWidth && xSides[0].scrollWidth === xSides[0].clientWidth) {
        ySides[0].setAttribute('style', 'padding-bottom: 10px');
      }

      if (xSides[0].scrollWidth > xSides[0].clientWidth && xSides[1].scrollWidth === xSides[1].clientWidth) {
        ySides[1].setAttribute('style', 'padding-bottom: 10px');
      }
    }
  }, [ref, props.maxHeight]);

  return (
    <div>
      <div className={styles.tips}>
        <Icon type="warning-circle" customclass={styles['warning-circle']} />
        <FormattedMessage id="common.diff.tips" />
      </div>
      <div
        style={{
          height: props.height ?? 'unset',
          maxHeight: props.maxHeight ?? 'unset',
        }}
        ref={ref}
        className="diff"
        dangerouslySetInnerHTML={{ __html: props.content }}
      />
    </div>
  );
};

export default Diff;
