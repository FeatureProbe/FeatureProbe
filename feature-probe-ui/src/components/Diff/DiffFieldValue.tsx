import { CSSProperties } from 'react';
import { diffType } from './constants';
import styles from './DiffFieldValue.module.scss';

export const DiffFieldValue = (props: { type?: diffType; value?: string | string[]; style?: CSSProperties }) => {
  const { type, style, value } = props;
  const typeCls = styles[`field-value-${type}`];

  if (value instanceof Array) {
    return (
      <div>
        {value.map((item) => {
          return <div style={style} className={`${styles['diff-field-value']} ${typeCls}`}>
            {item}
          </div>;
        })}
      </div>
    );
  }

  return (
    <div style={style} className={`${styles['diff-field-value']} ${typeCls}`}>
      {value}
    </div>
  );
};
