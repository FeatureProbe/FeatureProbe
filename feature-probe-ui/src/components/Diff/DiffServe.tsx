import { ArrayChange } from 'diff';
import { Table } from 'semantic-ui-react';
import { ArrayObj, DiffResult } from './diff';
import { DiffFieldValue } from './DiffFieldValue';
import { renderField, renderFieldsItems } from './renderDiff';
import conditionStyles from './RulesDiffContent.module.scss';
import styles from './DiffServe.module.scss';
import fieldStyles from './fields.module.scss';
import { FormattedMessage } from 'react-intl';

interface DiffServeContentProps {
  map?: Map<string, string>;
  content:
    | {
        split?: number[];
        select?: string;
      }
    | ArrayChange<ArrayObj>[]
    | ArrayChange<number>[];
  type: 'after' | 'before';
  diffType: 'remove' | 'same' | 'modify' | 'add';
}

export const DiffServeContent: React.FC<DiffServeContentProps> = (props) => {
  const { content, type, diffType } = props;
  return (
    <Table.Row className={`${fieldStyles[`diff-item-${diffType}`]} ${conditionStyles['condition-diff-item']}`}>
      <Table.Cell><FormattedMessage id='common.serve.text' /></Table.Cell>
      <Table.Cell colSpan="5">
        {(() => {
          if (content instanceof Array) {
            return renderField(content, type, (value) => {
              if (value.get('select')) {
                const select = value.get('select');
                if (select?.type === 'add' && type === 'after') {
                  return <DiffFieldValue type="add" value={select.value as string} />;
                } else if (select?.type === 'remove' && type === 'before') {
                  return <DiffFieldValue type="remove" value={select.value as string} />;
                } else if (select?.type === 'same') {
                  return <DiffFieldValue type="same" value={select.value as string} />;
                }
              }
              if (value.get('split')) {
                const split = value.get('split');
                if (split && split.value instanceof Array && typeof split.value[0] !== 'object') {
                  return <DiffFieldValue type={split.type} value={split.value} />;
                } else {
                  return renderFieldsItems(split?.value as ArrayChange<ArrayObj>[], type, (value, diffType) => {
                    if (diffType === 'add' && type === 'after') {
                      return <DiffFieldValue type="add" value={value as string} />;
                    } else if (diffType === 'remove' && type === 'before') {
                      return <DiffFieldValue type="remove" value={value as string} />;
                    } else if (diffType === 'same') {
                      return (
                        <DiffFieldValue
                          style={{
                            background: 'rgba(33,37,41,0.08)',
                          }}
                          type="same"
                          value={value as string}
                        />
                      );
                    }
                  });
                }
              }
            });
          } else {
            if (content.select !== undefined) {
              return <DiffFieldValue value={content.select} />;
            }
            if (content.split !== undefined) {
              return <DiffFieldValue value={content.split as unknown as string[]} />;
            }
          }
        })()}
      </Table.Cell>
    </Table.Row>
  );
};

interface DiffServeProps {
  content: DiffResult;
}

export const DiffServe: React.FC<DiffServeProps> = (props) => {
  const { content } = props;

  return (
    <div className={styles['serve-diff']}>
      <DiffServeContent diffType="modify" content={content as ArrayChange<ArrayObj>[]} type={'before'} />
      <DiffServeContent diffType="modify" content={content as ArrayChange<ArrayObj>[]} type={'after'} />
    </div>
  );
};
