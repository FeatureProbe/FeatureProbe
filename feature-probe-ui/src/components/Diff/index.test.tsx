import { render } from '@testing-library/react';
import JSONbig from 'json-bigint';
import { createPatch } from 'diff';
import { IntlWrapper } from 'components/utils/wrapper';
import { html } from 'diff2html/lib/diff2html';
import Diff from '.';
import '@testing-library/jest-dom';


it('Diff snapshot', (done) => {
  (async () => {
    const before = JSONbig.stringify({}, null, 2);
    const after = JSONbig.stringify({ name: 'test' }, null, 2);
    const result = createPatch('content', before.replace(/\\n/g, '\n'), after.replace(/\\n/g, '\n'));
    const content = html(result, {
      matching: 'lines',
      outputFormat: 'side-by-side',
      diffStyle: 'word',
      drawFileList: false,
    });
    const { asFragment } = render(<Diff content={content} />, {
      wrapper: IntlWrapper,
    });

    expect(asFragment()).toMatchSnapshot();

    done();
  })();
});

it('Diff opt', (done) => {
  (async () => {
    const before = JSONbig.stringify({}, null, 2);
    const after = JSONbig.stringify({ name: 'test' }, null, 2);
    const result = createPatch('content', before.replace(/\\n/g, '\n'), after.replace(/\\n/g, '\n'));
    const content = html(result, {
      matching: 'lines',
      outputFormat: 'side-by-side',
      diffStyle: 'word',
      drawFileList: false,
    });
    const { rerender } = render(<Diff key={1} maxHeight={500} height={500} content={content} />, {
      wrapper: IntlWrapper
    });
    const diffSides = document.getElementsByClassName('d2h-file-side-diff');
    diffSides[0].setAttribute('style', 'height: 600px;');
    diffSides[1].setAttribute('style', 'height: 600px;');

    diffSides[0].dispatchEvent(new Event('scroll'));
    diffSides[1].dispatchEvent(new Event('scroll'));

    let xSides = document.getElementsByClassName('d2h-code-wrapper');
    Object.defineProperty(xSides[0], 'clientWidth', {
      value: 500,
    });
    Object.defineProperty(xSides[0], 'scrollWidth', {
      value: 600,
    });
    rerender(<Diff key={1} maxHeight={510} height={500} content={content} />);

    rerender(<Diff key={2} maxHeight={510} height={500} content={content} />);
    xSides = document.getElementsByClassName('d2h-code-wrapper');
    Object.defineProperty(xSides[1], 'clientWidth', {
      value: 500,
    });
    Object.defineProperty(xSides[1], 'scrollWidth', {
      value: 600,
    });
    rerender(<Diff key={2} maxHeight={520} height={500} content={content} />);

    Object.defineProperty(xSides[0], 'clientWidth', {
      value: 500,
    });
    Object.defineProperty(xSides[0], 'scrollWidth', {
      value: 600,
    });
    rerender(<Diff key={2} maxHeight={530} height={500} content={content} />);

    done();
  })();
});
