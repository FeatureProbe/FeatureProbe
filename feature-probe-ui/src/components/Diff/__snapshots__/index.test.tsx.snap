// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Diff snapshot 1`] = `
<DocumentFragment>
  <div
    class="box"
  >
    <div
      class="tips"
    >
      <span
        class="iconfont icon-warning-circle warning-circle"
        customclass="warning-circle"
        type="warning-circle"
      />
      The following is a before and after comparison of this change and is for demonstration purposes only, no other action is required.
    </div>
    <div
      class="diff"
      style="max-height: unset;"
    >
      <div
        class="d2h-wrapper"
      >
        
    
        <div
          class="d2h-file-wrapper"
          data-lang=""
          id="d2h-530617"
        >
          
    
          <div
            class="d2h-file-header"
          >
            
      
            <span
              class="d2h-file-name-wrapper"
            >
              
    
              <svg
                aria-hidden="true"
                class="d2h-icon"
                height="16"
                version="1.1"
                viewBox="0 0 12 16"
                width="12"
              >
                
        
                <path
                  d="M6 5H2v-1h4v1zM2 8h7v-1H2v1z m0 2h7v-1H2v1z m0 2h7v-1H2v1z m10-7.5v9.5c0 0.55-0.45 1-1 1H1c-0.55 0-1-0.45-1-1V2c0-0.55 0.45-1 1-1h7.5l3.5 3.5z m-1 0.5L8 2H1v12h10V5z"
                />
                
    
              </svg>
                  
              <span
                class="d2h-file-name"
              >
                content
              </span>
              
    
              <span
                class="d2h-tag d2h-changed d2h-changed-tag"
              >
                CHANGED
              </span>
            </span>
            

            <label
              class="d2h-file-collapse"
            >
              
    
              <input
                class="d2h-file-collapse-input"
                name="viewed"
                type="checkbox"
                value="viewed"
              />
              
    Viewed

            </label>
            
    
          </div>
          
    
          <div
            class="d2h-files-diff"
          >
            
        
            <div
              class="d2h-file-side-diff"
            >
              
            
              <div
                class="d2h-code-wrapper"
              >
                
                
                <table
                  class="d2h-diff-table"
                >
                  
                    
                  <tbody
                    class="d2h-diff-tbody"
                  >
                    
                    
                    <tr>
                      
    
                      <td
                        class="d2h-code-side-linenumber d2h-info"
                      />
                      
    
                      <td
                        class="d2h-info"
                      >
                        
        
                        <div
                          class="d2h-code-side-line"
                        >
                          @@ -1,1 +1,3 @@
                        </div>
                        
    
                      </td>
                      

                    </tr>
                    <tr>
                      
    
                      <td
                        class="d2h-code-side-linenumber d2h-del d2h-change"
                      >
                        
      1
    
                      </td>
                      
    
                      <td
                        class="d2h-del d2h-change"
                      >
                        
        
                        <div
                          class="d2h-code-side-line"
                        >
                          
            
                          <span
                            class="d2h-code-line-prefix"
                          >
                            -
                          </span>
                          
            
                          <span
                            class="d2h-code-line-ctn"
                          >
                            {
                            <del>
                              }
                            </del>
                          </span>
                          
        
                        </div>
                        
    
                      </td>
                      

                    </tr>
                    <tr>
                      
    
                      <td
                        class="d2h-code-side-linenumber d2h-code-side-emptyplaceholder d2h-cntx d2h-emptyplaceholder"
                      >
                        
      
    
                      </td>
                      
    
                      <td
                        class="d2h-cntx d2h-emptyplaceholder"
                      >
                        
        
                        <div
                          class="d2h-code-side-line d2h-code-side-emptyplaceholder"
                        >
                          
            
                          <span
                            class="d2h-code-line-prefix"
                          >
                             
                          </span>
                          
            
                          <span
                            class="d2h-code-line-ctn"
                          >
                            <br />
                          </span>
                          
        
                        </div>
                        
    
                      </td>
                      

                    </tr>
                    <tr>
                      
    
                      <td
                        class="d2h-code-side-linenumber d2h-code-side-emptyplaceholder d2h-cntx d2h-emptyplaceholder"
                      >
                        
      
    
                      </td>
                      
    
                      <td
                        class="d2h-cntx d2h-emptyplaceholder"
                      >
                        
        
                        <div
                          class="d2h-code-side-line d2h-code-side-emptyplaceholder"
                        >
                          
            
                          <span
                            class="d2h-code-line-prefix"
                          >
                             
                          </span>
                          
            
                          <span
                            class="d2h-code-line-ctn"
                          >
                            <br />
                          </span>
                          
        
                        </div>
                        
    
                      </td>
                      

                    </tr>
                    
                    
                  </tbody>
                  
                
                </table>
                
            
              </div>
              
        
            </div>
            
        
            <div
              class="d2h-file-side-diff"
            >
              
            
              <div
                class="d2h-code-wrapper"
              >
                
                
                <table
                  class="d2h-diff-table"
                >
                  
                    
                  <tbody
                    class="d2h-diff-tbody"
                  >
                    
                    
                    <tr>
                      
    
                      <td
                        class="d2h-code-side-linenumber d2h-info"
                      />
                      
    
                      <td
                        class="d2h-info"
                      >
                        
        
                        <div
                          class="d2h-code-side-line"
                        />
                        
    
                      </td>
                      

                    </tr>
                    <tr>
                      
    
                      <td
                        class="d2h-code-side-linenumber d2h-ins d2h-change"
                      >
                        
      1
    
                      </td>
                      
    
                      <td
                        class="d2h-ins d2h-change"
                      >
                        
        
                        <div
                          class="d2h-code-side-line"
                        >
                          
            
                          <span
                            class="d2h-code-line-prefix"
                          >
                            +
                          </span>
                          
            
                          <span
                            class="d2h-code-line-ctn"
                          >
                            {
                          </span>
                          
        
                        </div>
                        
    
                      </td>
                      

                    </tr>
                    <tr>
                      
    
                      <td
                        class="d2h-code-side-linenumber d2h-ins"
                      >
                        
      2
    
                      </td>
                      
    
                      <td
                        class="d2h-ins"
                      >
                        
        
                        <div
                          class="d2h-code-side-line"
                        >
                          
            
                          <span
                            class="d2h-code-line-prefix"
                          >
                            +
                          </span>
                          
            
                          <span
                            class="d2h-code-line-ctn"
                          >
                              "name": "test"
                          </span>
                          
        
                        </div>
                        
    
                      </td>
                      

                    </tr>
                    <tr>
                      
    
                      <td
                        class="d2h-code-side-linenumber d2h-ins"
                      >
                        
      3
    
                      </td>
                      
    
                      <td
                        class="d2h-ins"
                      >
                        
        
                        <div
                          class="d2h-code-side-line"
                        >
                          
            
                          <span
                            class="d2h-code-line-prefix"
                          >
                            +
                          </span>
                          
            
                          <span
                            class="d2h-code-line-ctn"
                          >
                            }
                          </span>
                          
        
                        </div>
                        
    
                      </td>
                      

                    </tr>
                    
                    
                  </tbody>
                  
                
                </table>
                
            
              </div>
              
        
            </div>
            
    
          </div>
          

        </div>
        

      </div>
    </div>
  </div>
</DocumentFragment>
`;
